crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -12,94 +12,139 @@ pub use bevy_ecs_macros::WorldQuery;
+/// - **[`Entity`].**
+///   Gets the identifier of the queried entity.
+/// - **[`Option`].**
+///   By default, a world query only tests entities that are compatible with its type.
+///   Wrapping it into an `Option` will increase the query search space, and it will return `None` if an entity doesn't satisfy the `WorldQuery`.
+/// - **[`AnyOf`].**
+///   Equivalent to wrapping each world query inside it into an `Option`.
+/// - **[`ChangeTrackers`].**
+///   Similar to change detection filters but it is used as a query fetch parameter.
+///   It exposes methods to check for changes to the wrapped component.
 ///
 /// Implementing the trait manually can allow for a fundamentally new type of behaviour.
 ///
-/// The derive macro implements [`WorldQuery`] for your type and declares an additional struct
-/// which will be used as an item for query iterators. The implementation also generates two other
-/// structs that are used as [`WorldQuery::Fetch`](WorldQueryGats::Fetch) and
-/// [`WorldQuery::State`] associated types.
+/// # Trait derivation
+///
+/// Query design can be easily structured by deriving `WorldQuery` for custom types.
+/// Despite the added complexity, this approach has several advantages over using `WorldQuery` tuples.
+/// The most relevant improvements are:
 ///
-/// The derive macro requires every struct field to implement the `WorldQuery` trait.
+/// - Reusability across multiple systems.
+/// - There is no need to destructure a tuple since all fields are named.
+/// - Subqueries can be composed together to create a more complex query.
+/// - Methods can be implemented for the query items.
+/// - There is no hardcoded limit on the number of elements.
 ///
-/// **Note:** currently, the macro only supports named structs.
+/// This trait can only be derived if each field of the struct also implements `WorldQuery`.
+/// The derive macro only supports regular structs (structs with named fields).
 ///
 /// ```
 /// # use bevy_ecs::prelude::*;
 /// use bevy_ecs::query::WorldQuery;
-///
-/// #[derive(Component)]
-/// struct Foo;
-/// #[derive(Component)]
-/// struct Bar;
+/// #
+/// # #[derive(Component)]
+/// # struct ComponentA;
+/// # #[derive(Component)]
+/// # struct ComponentB;
 ///
 /// #[derive(WorldQuery)]
 /// struct MyQuery {
 ///     entity: Entity,
-///     // We must explicitly list out all lifetimes, as we are defining a struct
-///     foo: &'static Foo,
-///     bar: Option<&'static Bar>,
+///     // It is required that all reference lifetimes are explicitly annotated, just like in any
+///     // struct. Each lifetime should be 'static.
+///     component_a: &'static ComponentA,
+///     component_b: &'static ComponentB,
 /// }
 ///
 /// fn my_system(query: Query<MyQuery>) {
-///     for q in &query {
-///         // Note the type of the returned item.
-///         let q: MyQueryItem<'_> = q;
-///         q.foo;
+///     for q in query.iter() {
```

Comment:
```suggestion
///     for q in &query {
```
I think `IntoIterator` is [the default way](https://github.com/bevyengine/bevy/pull/5269) of Bevy doing it.

Reply:
Still not a fan of it, but I will get used.